node/src/components/rpc_server/rpcs.rs

```diff
@@ -209,65 +172,276 @@ pub trait RpcWithOptionalParams {
-    /// Creates the warp filter for this particular RPC.
-    fn create_filter<REv: ReactorEventT>(
+    /// Tries to parse the incoming JSON-RPC request's "params" field as
+    /// `Option<OptionalRequestParams>`.
+    fn try_parse_params(
+        maybe_params: Option<Value>,
+    ) -> Result<Option<Self::OptionalRequestParams>, Error> {
+        serde_json::from_value::<Option<Self::OptionalRequestParams>>(
+            maybe_params.unwrap_or(Value::Null),
+        )
+        .map_err(|error| {
+            Error::new(
+                ReservedErrorCode::InvalidParams,
+                format!("Failed to parse 'params' field: {}", error),
+            )
+        })
+    }
+
+    /// Registers this RPC as the handler for JSON-RPC requests whose "method" field is the same as
+    /// `Self::METHOD`.
+    fn register_as_handler<REv: ReactorEventT>(
         effect_builder: EffectBuilder<REv>,
         api_version: ProtocolVersion,
-    ) -> BoxedFilter<(Response<Body>,)> {
-        let with_params = warp::path(RPC_API_PATH)
-            .and(filters::json_rpc())
-            .and(filters::method(Self::METHOD))
-            .and(filters::params::<Self::OptionalRequestParams>())
-            .and_then(
-                move |response_builder: Builder, params: Self::OptionalRequestParams| {
-                    Self::handle_request(
-                        effect_builder,
-                        response_builder,
-                        Some(params),
-                        api_version,
-                    )
-                    .map_err(reject::custom)
-                },
-            );
-        let with_invalid_params = warp::path(RPC_API_PATH)
-            .and(filters::json_rpc())
-            .and(filters::method(Self::METHOD))
-            .and(filters::params::<Value>())
-            .and_then(
-                move |response_builder: Builder, _params: Value| async move {
-                    response_builder
-                        .error(warp_json_rpc::Error::INVALID_PARAMS)
-                        .map_err(|_| reject::reject())
-                },
-            );
-        let without_params = warp::path(RPC_API_PATH)
-            .and(filters::json_rpc())
-            .and(filters::method(Self::METHOD))
-            .and_then(move |response_builder: Builder| {
-                Self::handle_request(effect_builder, response_builder, None, api_version)
-                    .map_err(reject::custom)
-            });
-        with_params
-            .or(without_params)
-            .unify()
-            .or(with_invalid_params)
-            .unify()
-            .boxed()
+        handlers_builder: &mut RequestHandlersBuilder,
+    ) {
+        let handler = move |maybe_params| async move {
+            let params = Self::try_parse_params(maybe_params)?;
+            Self::do_handle_request(effect_builder, api_version, params).await
+        };
+        handlers_builder.register_handler(Self::METHOD, Arc::new(handler))
```

Comment:
if I'm not mistaken this function is common accross all traits - meaning there _could_ be a trait they all have to implement that has this implementation as a deafult.

Reply:
The `register_as_handler` is slightly different across the three traits, so I'm not sure it's worthwhile?  We'd still have to implement the function in these three different ways.